aten/src/ATen/native/quantized/cpu/q_avgpool.cpp

```diff
@@ -0,0 +1,202 @@
+        input_p + b * istrideB,
+        output_p + b * sizeD * osizeH * osizeW,
+        sizeD,
+        isizeH, isizeW,
+        osizeH, osizeW,
+        istrideD,
+        istrideH, istrideW);
+    }
+  });
+}
+
+void adaptive_avg_pool2d_out_template(Tensor& output, Tensor input,
+                                      IntArrayRef output_size) {
+  for (int64_t i = 0; i < input.dim(); i++) {
+    TORCH_CHECK(input.size(i) > 0,
+      "adaptive_avg_pooling2d(): expected input to have non-empty spatial "
+      "dimensions, but input has sizes ", input.sizes(), " with dimension ", i,
+      " being empty");
+  }
+
+  TORCH_CHECK((input.dim() == 3 || input.dim() == 4),
+    "non-empty 3D or 4D (batch mode) tensor expected for input");
+
+  /* sizes */
+  int64_t sizeD  = input.size(-3);
+  int64_t isizeH = input.size(-2);
+  int64_t isizeW = input.size(-1);
+  /* strides */
+  int64_t istrideD = input.stride(-3);
+  int64_t istrideH = input.stride(-2);
+  int64_t istrideW = input.stride(-1);
+
+  auto osizeH = output_size[0];
+  auto osizeW = output_size[1];
+
+  /* resize output */
+  std::vector<int64_t> output_sizes;
+  int64_t sizeB = 0;
+  if (input.dim() == 3) {
+    output_sizes = {sizeD, osizeH, osizeW};
+  } else {
+    sizeB = input.size(-4);
+    output_sizes = {sizeB, sizeD, osizeH, osizeW};
+  }
+  if (!output.is_quantized() || output.sizes() != output_sizes) {
+    output = at::_empty_affine_quantized(output_sizes, input.options(),
+                                         input.q_scale(), input.q_zero_point());
+  }
+
+  if (input.dim() == 3 || input.size(0) == 1) {
+    AT_DISPATCH_QINT_TYPES(input.scalar_type(),
+      "quantized_adaptive_avg_pool2d", [&] {
+        auto input_data = input.data<scalar_t>();
+        auto output_data = output.data<scalar_t>();
+        adaptive_avg_pool2d_single_out_frame<scalar_t, underlying_t>(
+          input_data,
+          output_data,
+          sizeD,
+          isizeH, isizeW,
+          osizeH, osizeW,
+          istrideD,
+          istrideH, istrideW);
+      }
+    );
+  } else {
+    int64_t istrideB = input.stride(-4);
+
+    AT_DISPATCH_QINT_TYPES(input.scalar_type(),
+      "quantized_adaptive_avg_pool2d", [&] {
+        auto input_data = input.data<scalar_t>();
+        auto output_data = output.data<scalar_t>();
+        adaptive_avg_pool2d_out_frame<scalar_t, underlying_t>(
+          input_data,
+          output_data,
+          sizeB,
+          sizeD,
+          isizeH, isizeW,
+          osizeH, osizeW,
+          istrideB,
+          istrideD,
+          istrideH, istrideW);
+      }
+    );
+  }
+}
+
+class QAdaptiveAvgPool2D final : public torch::OperatorKernel {
+ public:
+  Tensor operator()(Tensor qx, std::vector<int64_t> output_size) {
+    return at::adaptive_avg_pool2d(qx, output_size);
+  }
+};
+
+static auto registry = torch::RegisterOperators().op(
+  "quantized::adaptive_avg_pool2d(Tensor qx, int[] output_size) -> Tensor",
```

Comment:
since you already use native_functions, I don't think you need to register it again

Reply:
I agree -- we were "double-registering" all the native_functions. I wanted to clean them up at some point :)